lib/adapters/parsers/grok.js

```diff
@@ -0,0 +1,76 @@
+var base = require('./base');
+var errors = require('../../errors');
+var grok = require('node-grok');
+var Promise = require('bluebird');
+var readlines = require('readlines');
+
+module.exports = base.extend({
+
+    initialize: function(options) {
+        this.pattern = options.pattern || '%{GREEDYDATA:message}';
+    },
+
+    parseStream: function(stream, emit) {
+        var self = this;
+
+        return new Promise(function(resolve, reject) {
+            grok.loadDefault(function(patterns) {
+                var buffer = [];
+                var liner = readlines.readlineStream();
+                var pattern = patterns.createPattern(self.pattern);
+
+                // because of the asynchronous way that patterns parse we need
+                // to track the pending parses and when the stream has ended to
+                // do the right thing.
+                var pending = 0;
+                var streamEnd = false;
+
+                liner.on('readable', function(){
+                    var line;
+                    while ((line = liner.read())) {
+                        line = line.toString();
+                        pending++;
+                        pattern.parse(line, function (err, point) {
+                            if (err) {
+                                reject(errors.runtimeError('RT-INTERNAL-ERROR',{
+                                    detail: err.toString()
+                                }));
+                            } else {
+                                buffer.push(point);
+
+                                if (buffer.length >= self.limit) {
+                                    emit(buffer);
+                                    buffer = [];
+                                }
+                            }
+
+                            if (streamEnd) {
```

Comment:
This seems incorrect -- what if pending is > 0 at this point but you've gotten an eof?

Overall it seems like you should DRY up the code by adding a `checkDone(pending, streamEnd)` function that you could call both here and in the 'end' event handler. In fact that could also do the check to emit when the buffer exceeds the limit as well which should tighten things up as well.